G+Smo Adapter Documentation on preCICE Website

Crazy-Rich-Meghan opened the pull request:
Hello,

This PR is the documentation for the G+Smo adapter.

Crazy-Rich-Meghan committed:
Update G+Smo information in adapter-overview.md

Crazy-Rich-Meghan committed:
Create adapter-gismo-get-adapter.md

Crazy-Rich-Meghan committed:
Create adapter-gismo-overview.md

Crazy-Rich-Meghan committed:
Update adapter documentation

hverhelst commented on pages/docs/adapters/gismo/adapter-gismo-get-adapter.md:
````diff
@@ -0,0 +1,23 @@
+cd gismo/build
+cmake .. -DGISMO_OPTIONAL="<Other submodules>;gsPreCICE" 
+```
+Build and install the tutorial:
````
```suggestion
Build and install the tutorial, to make it discoverable:
```

Crazy-Rich-Meghan committed:
Update pages/docs/adapters/gismo/adapter-gismo-get-adapter.md

Co-authored-by: Hugo Verhelst <38376601+hverhelst@users.noreply.github.com>

uekerman commented on pages/docs/adapters/adapter-overview.md:
```diff
@@ -39,7 +39,7 @@ Wherever meaningful (license, maturity of the project, no other home), we host t
 | [FEAP](http://projects.ce.berkeley.edu/feap/) | [TU Darmstadt FNB](https://www.fnb.tu-darmstadt.de/) | None | Structure part in FSI | |
 | [FEniCS-X](https://fenicsproject.org/) | [Benjamin Rodenberg, TUM](https://www.in.tum.de/i05/personen/personen/benjamin-rodenberg/) | [code](https://github.com/precice/fenicsx-adapter) | Structure part in CHT, FSI, any FEM | looking for maintainers! |
-| [G+Smo](https://gismo.github.io/) | [TU Delft Numerical Analysis](https://www.tudelft.nl/en/eemcs/the-faculty/departments/applied-mathematics/numerical-analysis/) | [code](https://github.com/gismo/gismo/tree/stable/extensions/gsPreCICE) | Structure part in CHT ||
+| [G+Smo](https://gismo.github.io/) | [TU Delft Numerical Analysis](https://www.tudelft.nl/en/eemcs/the-faculty/departments/applied-mathematics/numerical-analysis/) |[code](https://github.com/gismo/gsPreCICE) | Structure part in CHT; Structure part in FSI |Based on IsoGeometric Analysis|
```
```suggestion
| [G+Smo](https://gismo.github.io/) | [TU Delft Numerical Analysis](https://www.tudelft.nl/en/eemcs/the-faculty/departments/applied-mathematics/numerical-analysis/) |[code](https://github.com/gismo/gsPreCICE) | Structure part in CHT; Structure part in FSI ||
```
That's not the right place for the this information. We mainly use the comments column for the status of the project.

uekerman commented on pages/docs/adapters/gismo/adapter-gismo-get-adapter.md:
```diff
@@ -0,0 +1,23 @@
+
+This adapter provides a collection of examples demonstrating the use of a G+Smo solver adapted for preCICE. 
+
+After downloading the G+Smo main library, clone the G+Smo adapter submodule:
```
```suggestion
After downloading [G+Smo](https://github.com/gismo/gismo), clone the G+Smo adapter submodule:
```
Do the following lines really clone the adapter (implicitly)?

uekerman commented on pages/docs/adapters/gismo/adapter-gismo-get-adapter.md:
````diff
@@ -0,0 +1,23 @@
+cd gismo/build
+cmake .. -DGISMO_OPTIONAL="<Other submodules>;gsPreCICE" 
+```
+Build and install the tutorial, to make it discoverable:
````
```suggestion

Build and install the tutorial, to make it discoverable:
```
I think we need to explain better what you mean by tutorial here. This is sth I need to run in the `gismo` subfolder of a preCICE tutorial? Then, we should this there.

uekerman commented on pages/docs/adapters/gismo/adapter-gismo-overview.md:
```diff
@@ -0,0 +1,21 @@
+---
+title: The G+Smo adapter
+permalink: adapter-gismo-overview.html
+redirect_from: adapter-gismo.html
```
```suggestion
```
Sth like this is only necessary for backwards compatibility.

uekerman commented on pages/docs/adapters/gismo/adapter-gismo-overview.md:
```diff
@@ -0,0 +1,21 @@
+title: The G+Smo adapter
+permalink: adapter-gismo-overview.html
+redirect_from: adapter-gismo.html
+keywords: adapter, G+Smo, IsoGeometric Analysis
```
```suggestion
keywords: adapter, G+Smo, Isogeometric Analysis
```

uekerman commented on pages/docs/adapters/gismo/adapter-gismo-overview.md:
```diff
@@ -0,0 +1,21 @@
+---
+
+## What is G+Smo?
+G+Smo (pronounced gismo or gizmo) is a C++ library for isogeometric analysis (IGA). Geometry plus simulation modules aims at the seamless integration of Computer-aided Design (CAD) and Finite Element Analysis (FEA).
```
```suggestion

G+Smo (pronounced gismo or gizmo) is a C++ library for isogeometric analysis (IGA). Geometry plus simulation modules aims at the seamless integration of Computer-aided Design (CAD) and Finite Element Analysis (FEA).
```

uekerman commented on pages/docs/adapters/gismo/adapter-gismo-overview.md:
```diff
@@ -0,0 +1,21 @@
+
+## Start here
+1. [Get G+Smo and the adapter](adapter-gismo-get-adapter.html)
+2. [Configure and run tutorial](adapter-gismo-get-adapter.html).
```
Let's try to remove some duplication. Short pages are good.

uekerman reviewed (commented):
Thanks for the contribution here! Let me first try out the tutorials and iterate there.  Once the tutorials are merged, let's continue here.

You also have some documentation here: https://github.com/gismo/gsPreCICE
It could be good to reduce such duplication from the beginning. You could create a `docs` folder in your repo and move the pages from this PR to there. Then, we include the adapter is a submodule in the website repo. This is also how we handle other adapter docs (e.g. the [OpenFOAM adapter docs](https://github.com/precice/openfoam-adapter/tree/develop/docs)). I am happy to help with this.

Crazy-Rich-Meghan committed:
Update pages/docs/adapters/gismo/adapter-gismo-overview.md

Co-authored-by: Benjamin Uekermann <benjamin.uekermann@gmail.com>

Crazy-Rich-Meghan committed:
Update pages/docs/adapters/gismo/adapter-gismo-overview.md

Co-authored-by: Benjamin Uekermann <benjamin.uekermann@gmail.com>

Crazy-Rich-Meghan committed:
Update pages/docs/adapters/adapter-overview.md

Co-authored-by: Benjamin Uekermann <benjamin.uekermann@gmail.com>

Crazy-Rich-Meghan committed:
Update pages/docs/adapters/gismo/adapter-gismo-overview.md

Co-authored-by: Benjamin Uekermann <benjamin.uekermann@gmail.com>

Crazy-Rich-Meghan committed:
Update pages/docs/adapters/gismo/adapter-gismo-get-adapter.md

Co-authored-by: Benjamin Uekermann <benjamin.uekermann@gmail.com>

Crazy-Rich-Meghan committed:
Update pages/docs/adapters/gismo/adapter-gismo-get-adapter.md

Co-authored-by: Benjamin Uekermann <benjamin.uekermann@gmail.com>

Crazy-Rich-Meghan commented:
> Thanks for the contribution here! Let me first try out the tutorials and iterate there. Once the tutorials are merged, let's continue here.
> 
> You also have some documentation here: https://github.com/gismo/gsPreCICE It could be good to reduce such duplication from the beginning. You could create a `docs` folder in your repo and move the pages from this PR to there. Then, we include the adapter is a submodule in the website repo. This is also how we handle other adapter docs (e.g. the [OpenFOAM adapter docs](https://github.com/precice/openfoam-adapter/tree/develop/docs)). I am happy to help with this.

Hi @uekerman,

For the documentation from our side, there will be some duplications anyway. Since G+Smo is using the Doxygen system which uses `/doc`. Would you still recommend to have `/docs` for G+Smo and `/docs` for preCICE in our submodule instead of having it here?

Crazy-Rich-Meghan committed:
Create adapter-gismo.md

Crazy-Rich-Meghan committed:
Update adapter-gismo.md

Crazy-Rich-Meghan committed:
Merge branch 'precice:master' into master

Crazy-Rich-Meghan committed:
Delete pages/docs/adapters/gismo directory

Crazy-Rich-Meghan commented:
@uekerman we removed the G+Smo documentation folder. And added everything into a single page, the page name is `adapter-gismo.md`.

uekerman committed:
Tweak README, add to sidebar

uekerman commented:
I modified the README and the sidebar (locally). Could you please edit the PR setup and tick "allow edits by maintainers". Then, I could push my changes to your master branch. Or you could add me as collaborator to your fork and give write access.

Crazy-Rich-Meghan committed:
Merge branch 'master' into master

Crazy-Rich-Meghan commented:
> I modified the README and the sidebar (locally). Could you please edit the PR setup and tick "allow edits by maintainers". Then, I could push my changes to your master branch. Or you could add me as collaborator to your fork and give write access.

@uekerman I added you as the collaborator of my fork.

uekerman committed:
Merge pull request #1 from Crazy-Rich-Meghan/gismo_master

Tweak README, add to sidebar

Crazy-Rich-Meghan committed:
Added solver dependency

uekerman commented on pages/docs/adapters/adapter-gismo.md:
```diff
@@ -0,0 +1,42 @@
+
+## Aim of the adapter
+
+The G#Smo adapter provides a collection of examples demonstrating the use of G+Smo solvers adapted for preCICE. A particular focus lies on the IGA functionality of G#Smo.
```
```suggestion
The G+Smo adapter provides a collection of examples demonstrating the use of G+Smo solvers adapted for preCICE. A particular focus lies on the IGA functionality of G+Smo.
```

uekerman commented on pages/docs/adapters/adapter-overview.md:
```diff
@@ -39,7 +39,7 @@ Wherever meaningful (license, maturity of the project, no other home), we host t
 | [FEAP](http://projects.ce.berkeley.edu/feap/) | [TU Darmstadt FNB](https://www.fnb.tu-darmstadt.de/) | None | Structure part in FSI | |
 | [FEniCS-X](https://fenicsproject.org/) | [Benjamin Rodenberg, TUM](https://www.in.tum.de/i05/personen/personen/benjamin-rodenberg/) | [code](https://github.com/precice/fenicsx-adapter) | Structure part in CHT, FSI, any FEM | looking for maintainers! |
-| [G+Smo](https://gismo.github.io/) | [TU Delft Numerical Analysis](https://www.tudelft.nl/en/eemcs/the-faculty/departments/applied-mathematics/numerical-analysis/) | [code](https://github.com/gismo/gismo/tree/stable/extensions/gsPreCICE) | Structure part in CHT ||
+| [G+Smo](https://gismo.github.io/) | [TU Delft Numerical Analysis](https://www.tudelft.nl/en/eemcs/the-faculty/departments/applied-mathematics/numerical-analysis/) |[code](https://github.com/gismo/gsPreCICE) | Structure part in CHT and FSI ||
```
```suggestion
| [G+Smo](https://gismo.github.io/) | [TU Delft Numerical Analysis](https://www.tudelft.nl/en/eemcs/the-faculty/departments/applied-mathematics/numerical-analysis/) |[code](https://github.com/gismo/gsPreCICE) | Structure part in FSI ||
```

uekerman committed:
Update pages/docs/adapters/adapter-overview.md

uekerman committed:
Update pages/docs/adapters/adapter-gismo.md

uekerman commented on pages/docs/adapters/adapter-gismo.md:
```diff
@@ -0,0 +1,42 @@
+
+| **Solver**                          | **Required Submodules**                                           | **Configuration**                                |
+|------------------------------------|-------------------------------------------------------------------|---------------------------------------------------|
+| **perpendicular-flap-stress** | [`gsElasticity`](https://github.com/gismo/gsElasticity),[`gsStructuralAnalysis`](https://github.com/gismo/gsStructuralAnalysis)   | `cmake .. -DGISMO_OPTIONAL="gsPreCICE;gsElasticity;gsStructuralAnalysis"`| 
```
```suggestion
| **perpendicular-flap-vertex-gismo** | [`gsElasticity`](https://github.com/gismo/gsElasticity),[`gsStructuralAnalysis`](https://github.com/gismo/gsStructuralAnalysis)   | `cmake .. -DGISMO_OPTIONAL="gsPreCICE;gsElasticity;gsStructuralAnalysis"`| 
```
@Crazy-Rich-Meghan right?

Crazy-Rich-Meghan committed:
Merge branch 'master' into master

Crazy-Rich-Meghan committed:
Update pages/docs/adapters/adapter-gismo.md

Co-authored-by: Benjamin Uekermann <benjamin.uekermann@gmail.com>

uekerman committed:
Fix markdown formatting

uekerman committed:
Fix markdown formatting

MakisH commented on pages/docs/adapters/adapter-gismo.md:
```diff
@@ -0,0 +1,42 @@
+
+| **Solver**                          | **Required Submodules**                                           | **Configuration**                                |
+|------------------------------------|-------------------------------------------------------------------|---------------------------------------------------|
+| **perpendicular-flap-vertex-gismo** | [`gsElasticity`](https://github.com/gismo/gsElasticity),[`gsStructuralAnalysis`](https://github.com/gismo/gsStructuralAnalysis)   | `cmake .. -DGISMO_OPTIONAL="gsPreCICE;gsElasticity;gsStructuralAnalysis"`|
```
Since this is something to copy from:

```suggestion
| `perpendicular-flap-vertex-gismo` | [`gsElasticity`](https://github.com/gismo/gsElasticity),[`gsStructuralAnalysis`](https://github.com/gismo/gsStructuralAnalysis)   | `cmake .. -DGISMO_OPTIONAL="gsPreCICE;gsElasticity;gsStructuralAnalysis"`|
```

MakisH commented on _data/sidebars/docs_sidebar.yml:
```diff
@@ -396,6 +396,14 @@ entries:
           url: /adapter-fenics.html
           output: web, pdf
+          
+      - title: G+Smo
```
@uekerman I get the impression that we are not clear whether this is an "official" or "third-party" adapter here: on the one hand, we list it together with the official adapters in the sidebar, and host the documentation (which is great for discovering it). On the other hand, it is hosted under `github.com/gismo` and we still list it as a third-party adapter in the respective table.

Maybe the clearest way forward would be to organize the sidebar as:

```
- Official adapters
  - CalculiX
  - code_aster
- Third-party adapters
  - G+Smo
    - (Subpages of G+Smo) 
  - ... 
``` 

or, alternatively:

```
- Provided adapters
   - CalculiX
   - code_aster
   - ...
   - Third-party adapters
   - G+Smo 
```

but then we lose the subpages feature (I think). We could even go one step further and introduce `Legacy adapters`, reflecting the overview table.

One negative effect of this distinction would be that it would be more complicated to find the page you are looking for, if you don't know in which category it is.

Long-term: Maybe all the adapters pages could even form a new top-level section on the website (between `Docs` and `Tutorials`). It could even be named `Ecosystem` and include the application cases as well, with a schematic overview of components in the first page.

uekerman replied:
I see your points. I cheated a bit as we currently say "Provided adapters", so we don't distinguish between official and third-party.
I would postpone such a restructuring till we have to do it for preECO anyway. Related to https://github.com/precice/preeco-orga/issues/12
Then, we should probably also drop the name "official" and list maintainers instead.

MakisH replied:
OK, then let's postpone the restructuring.

MakisH commented on pages/docs/adapters/adapter-gismo.md:
I think still interesting here would be the following sections:

- Configuration
- Citing
- History / Related literature

as well as links to any further documentation

uekerman replied:
I added brief sections on how to cite G+Smo (linking to the [FAQ](https://github.com/gismo/gismo/wiki/Frequently-asked-questions#user-content-How_can_I_cite_GSmo_in_a_publicationpresentation_)) and on the current developers. @Crazy-Rich-Meghan Please protest if we should change anything here.

https://github.com/precice/precice.github.io/pull/482/commits/b228f9bf1d90b6ea708c63a8339a805df2d85c62

Configuration: does currently not exist, but plans are to work towards following the schema. I would suggest to add such a section once we are there.

MakisH commented on pages/docs/adapters/adapter-gismo.md:
````diff
@@ -0,0 +1,42 @@
+
+```bash
+make install <solver_name>
+```
````
A copy-paste example would at least save us maintainers some time whenever we test things (the rendering on GitHub is probably broken):

```suggestion
```

For example:

```bash
cmake .. -DGISMO_OPTIONAL="gsPreCICE;gsElasticity;gsStructuralAnalysis"
make -j 8 perpendicular-flap-vertex-gismo
make install perpendicular-flap-vertex-gismo
```
```
```

uekerman replied:
The vision is that more solvers will be added here. Isn't the code above enough for copy-paste:

```bash
git clone https://github.com/gismo/gismo.git
cd gismo
mkdir build & cd build
cmake .. -DGISMO_OPTIONAL="<Other submodules>;gsPreCICE"
make <solver_name>
```

I didn't test the rendering, maybe sth is broken there.

MakisH replied:
Well, it is enough but involves some additional step to read and figure out what needs to go into `<solver_name>` and `<Other submodules>`. Which is general documentation, but not a Quickstart for someone that does not yet know what they need. It is an overview reference, but not a concrete example.

By "rendering", I meant in my comment: how GitHub displays my "code suggestion" that includes a code snippet.

Edit: And yes, I clearly see the vision. But we also have a Quickstart tutorial in preCICE next to everything else.

uekerman replied:
I added an example.

MakisH commented on pages/docs/adapters/adapter-gismo.md:
```diff
@@ -0,0 +1,42 @@
+|------------------------------------|-------------------------------------------------------------------|---------------------------------------------------|
+| **perpendicular-flap-vertex-gismo** | [`gsElasticity`](https://github.com/gismo/gsElasticity),[`gsStructuralAnalysis`](https://github.com/gismo/gsStructuralAnalysis)   | `cmake .. -DGISMO_OPTIONAL="gsPreCICE;gsElasticity;gsStructuralAnalysis"`|
+
+Finally, make the solver discoverable, e.g. by installation:
```
```suggestion
Finally, make the solver discoverable, e.g. by installation (by default, to `/usr/local/`, which might not be in your `LD_LIBRARY_PATH`):
```

uekerman committed:
Update pages/docs/adapters/adapter-gismo.md

Co-authored-by: Gerasimos Chourdakis <gerasimos.chourdakis@ipvs.uni-stuttgart.de>

uekerman committed:
Update pages/docs/adapters/adapter-gismo.md

Co-authored-by: Gerasimos Chourdakis <gerasimos.chourdakis@ipvs.uni-stuttgart.de>

uekerman committed:
Add how to cite and history section

uekerman committed:
Add how to cite and history section

uekerman committed:
Add example instructions

uekerman commented on pages/docs/adapters/adapter-gismo.md:
```diff
@@ -0,0 +1,60 @@
+cd gismo
+mkdir build & cd build
+cmake .. -DGISMO_OPTIONAL="gsElasticity;gsStructuralAnalysis;gsPreCICE"
+make install <perpendicular-flap-vertex-gismo>
```
```suggestion
make install perpendicular-flap-vertex-gismo
```

uekerman committed:
Update pages/docs/adapters/adapter-gismo.md

MakisH reviewed (approved):
Looks good, feel free to merge, ideally once the tutorial is also ready to merge.
